File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/ui/fragment/dialog/IconPickerDialogFragment.java
package de.szalkowski.activitylauncher.ui.fragment.dialog;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.GridView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import de.szalkowski.activitylauncher.R;
import de.szalkowski.activitylauncher.async.AsyncProvider;
import de.szalkowski.activitylauncher.async.IconListAsyncProvider;
import de.szalkowski.activitylauncher.ui.adapter.IconListAdapter;

public class IconPickerDialogFragment extends DialogFragment implements IconListAsyncProvider.Listener<IconListAdapter> {
    private GridView grid;
    private IconPickerListener listener = null;

    @Override
    public void onAttach(@NonNull Context activity) {
        super.onAttach(activity);

        IconListAsyncProvider provider = new IconListAsyncProvider(this.getActivity(), this);
        provider.execute();
    }

    public void attachIconPickerListener(IconPickerListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(requireActivity());
        LayoutInflater inflater = (LayoutInflater) requireActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View view = inflater.inflate(R.layout.icon_picker, null);

        this.grid = (GridView) view;
        this.grid.setOnItemClickListener((view1, item, index, id) -> {
            if (IconPickerDialogFragment.this.listener != null) {
                IconPickerDialogFragment.this.listener.iconPicked(view1.getAdapter().getItem(index).toString());
                IconPickerDialogFragment.this.requireDialog().dismiss();
            }
        });

        builder.setTitle(R.string.title_dialog_icon_picker)
                .setView(view)
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> IconPickerDialogFragment.this.requireDialog().cancel());

        return builder.create();
    }

    @Override
    public void onProviderFinished(AsyncProvider<IconListAdapter> task,
                                   IconListAdapter value) {
        try {
            this.grid.setAdapter(value);
        } catch (Exception e) {
            Toast.makeText(this.getActivity(), R.string.error_icons, Toast.LENGTH_SHORT).show();
        }
    }

    public interface IconPickerListener {
        void iconPicked(String icon);
    }
}

File: app/src/main/java/de/szalkowski/activitylauncher/ui/IconPickerDialogFragment.kt
package de.szalkowski.activitylauncher.ui

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import dagger.hilt.android.AndroidEntryPoint
import de.szalkowski.activitylauncher.R
import de.szalkowski.activitylauncher.databinding.IconPickerBinding
import de.szalkowski.activitylauncher.services.IconLoaderService
import javax.inject.Inject

@AndroidEntryPoint
class IconPickerDialogFragment : DialogFragment(), AsyncProvider.Listener<IconListAdapter> {
    @Inject
    internal lateinit var iconListAsyncProviderFactory: IconListAsyncProvider.IconListAsyncProviderFactory

    private var listener: IconPickerListener? = null
    private var _binding: IconPickerBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onAttach(activity: Context) {
        super.onAttach(activity)

        val provider = iconListAsyncProviderFactory.create(this)
        provider.execute()
    }

    fun attachIconPickerListener(listener: IconPickerListener) {
        this.listener = listener
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(requireActivity())
        _binding = IconPickerBinding.inflate(layoutInflater, null, false)
        val view = _binding!!.root

        binding.gvIcons.onItemClickListener =
            OnItemClickListener { adapterView: AdapterView<*>, _: View?, index: Int, _: Long ->
                listener?.iconPicked(
                    // FIXME ugly and unsafe
                    (adapterView.adapter.getItem(index) as IconLoaderService.IconInfo).iconResourceName
                )
                requireDialog().dismiss()
            }

        builder.setTitle(R.string.title_dialog_icon_picker).setView(view).setNegativeButton(
            android.R.string.cancel
        ) { dialog: DialogInterface?, _: Int ->
            dialog?.cancel()
        }

        return builder.create()
    }

    override fun onProviderFinished(task: AsyncProvider<IconListAdapter>?, value: IconListAdapter) {
        try {
            binding.gvIcons.adapter = value
            binding.progressCircular.visibility = View.GONE
            binding.gvIcons.visibility = View.VISIBLE
        } catch (ignored: Exception) {
            Toast.makeText(this.activity, R.string.error_icons, Toast.LENGTH_SHORT).show()
        }
    }

    fun interface IconPickerListener {
        fun iconPicked(icon: String)
    }
}
